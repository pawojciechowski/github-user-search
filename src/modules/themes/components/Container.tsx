import styled from 'styled-components';
import { Theme } from '../types';

const Container = styled.div`
  padding-left: ${({ theme }: { theme: Theme }) => `${theme.spacer * 1.5}px`};
  padding-right: ${({ theme }: { theme: Theme }) => `${theme.spacer * 1.5}px`};

  @media screen and (min-width: 768px) {
    max-width: 700px;
    margin: auto;
  }
`;

export default Container;
