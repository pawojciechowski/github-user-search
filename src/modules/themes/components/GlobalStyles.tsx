import { baseFontCSS } from '../styles';
import { createGlobalStyle } from 'styled-components';
import { Theme } from '../types';

export default createGlobalStyle`
  html,
  body {
    height: 100%;
  }

  body {
    ${baseFontCSS}
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    margin: 0;
    padding: 0;
    background: ${({ theme }: { theme: Theme }) => theme.background};
  }

  #root {
    height: 100%;
  }

  img {
    max-width: 100%;
    display: block;
  }
`;
