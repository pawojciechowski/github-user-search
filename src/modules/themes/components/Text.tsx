import styled from 'styled-components';
import { Theme } from '../types';

const Text = styled.p`
  margin: 0;
  padding-bottom: ${({ theme }: { theme: Theme }) => `${theme.spacer}px`};
`;

export default Text;
