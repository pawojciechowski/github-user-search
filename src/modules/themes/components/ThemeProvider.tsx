import React, { PropsWithChildren } from 'react';
import { ThemeProvider as StyledComponentsThemeProvider, createGlobalStyle } from 'styled-components';
import { Theme, ThemeName } from '../types';
import { themes } from '../themes';
import { baseFontCSS } from '../styles';

const GlobalStyle = createGlobalStyle`
  html,
  body {
    height: 100%;
  }

  body {
    ${baseFontCSS}
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    margin: 0;
    padding: 0;
    background: ${({ theme }: { theme: Theme }) => theme.background};
  }

  #root {
    height: 100%;
  }

  img {
    max-width: 100%;
    display: block;
  }
`;

export function ThemeProvider({ children }: PropsWithChildren<{}>) {
  // TODO: Get theme from state
  const themeName: ThemeName = 'default';

  return (
    <StyledComponentsThemeProvider theme={themes[themeName]}>
      <GlobalStyle />
      {children}
    </StyledComponentsThemeProvider>
  );
}

export default ThemeProvider;
