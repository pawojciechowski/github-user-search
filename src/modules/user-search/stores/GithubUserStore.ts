import { types, flow, getEnv, Instance, getSnapshot } from 'mobx-state-tree';
import { IAPIService } from 'modules/api/types';
import { GithubUserResponse, GithubReposResponse } from '../../../config/api/types';
import { APIError } from 'modules/api/errors';

export const GithubRepo = types.model('GithubRepo', {
  id: types.identifierNumber,
  name: types.string,
  htmlUrl: types.string,
  stars: types.number,
});

export const GithubUser = types.model('GithubUser', {
  login: types.identifier,
  name: types.string,
  bio: types.maybeNull(types.string),
  avatarUrl: types.string,
  repos: types.array(GithubRepo),
});

export const GithubUserStore = types
  .model('GithubUserStore', {
    user: types.maybeNull(GithubUser),
    state: types.optional(
      types.union(
        types.literal('initial'),
        types.literal('pending'),
        types.literal('success'),
        types.literal('error'),
        types.literal('notFound')
      ),
      'initial'
    ),
    error: types.maybe(types.string),
    lastSearchUsername: types.maybe(types.string),
  })
  .actions((self) => ({
    fetchUserWithRepos: flow(function* fetchUserWithRepos(username) {
      self.user = null;
      self.lastSearchUsername = username;
      self.state = 'pending';

      const apiService: IAPIService = getEnv(self).apiService;
      const userResponsePromise = apiService.get<GithubUserResponse>(`/users/${username}`);
      const reposResponsePromise = apiService.get<GithubReposResponse>(`/users/${username}/repos`);

      try {
        const [userResponse, reposResponse]: [GithubUserResponse, GithubReposResponse] = yield Promise.all([
          userResponsePromise,
          reposResponsePromise,
        ]);
        self.user = GithubUser.create({
          login: userResponse.login,
          name: userResponse.name || '',
          bio: userResponse.bio,
          avatarUrl: userResponse.avatar_url,
          repos: reposResponse.map((repo) =>
            GithubRepo.create({
              id: repo.id,
              name: repo.name,
              htmlUrl: repo.html_url,
              stars: repo.stargazers_count,
            })
          ),
        });
        self.state = 'success';
      } catch (error) {
        if (error instanceof APIError && error.response?.status === 404) {
          self.state = 'notFound';
          self.user = null;
        } else {
          self.state = 'error';
          self.error = error.message;
          self.user = null;
        }
      }
    }),
  }))
  .views((self) => ({
    get userWithTopRepos() {
      const snapshot = getSnapshot(self);

      if (snapshot.user === null) {
        return null;
      } else {
        return GithubUser.create({
          ...snapshot.user,
          repos: [...snapshot.user.repos].sort((repo1, repo2) => repo2.stars - repo1.stars).slice(0, 3),
        });
      }
    },
    get isLoading() {
      return self.state === 'pending';
    },
  }));

export interface IGithubUserStore extends Instance<typeof GithubUserStore> {}
export interface IGithubUser extends Instance<typeof GithubUser> {}
