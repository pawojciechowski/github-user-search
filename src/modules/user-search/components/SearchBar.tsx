import React, { useState } from 'react';
import { observer, inject } from 'mobx-react';
import { IGithubUserStore } from '../stores/GithubUserStore';
import Button from 'common/button/components/Button';
import Input from 'common/input/components/Input';
import styled from 'styled-components';
import { Theme } from 'modules/themes/types';

const SearchBarContainer = styled.div`
  display: flex;

  ${Input} {
    flex-grow: 1;
    margin-right: ${({ theme }: { theme: Theme }) => `${theme.spacer}px`};
  }
`;

const handleKeyPress = (e: React.KeyboardEvent<HTMLInputElement>, userStore: IGithubUserStore) => {
  if (e.key === 'Enter') {
    userStore.fetchUserWithRepos(e.currentTarget.value);
  }
};

const handleChange = (
  e: React.FormEvent<HTMLInputElement>,
  setInputValue: React.Dispatch<React.SetStateAction<string>>
) => {
  setInputValue(e.currentTarget.value);
};

interface SearchBarProps {
  userStore: IGithubUserStore;
}

export function SearchBar({ userStore }: SearchBarProps) {
  const [value, setInputValue] = useState('');
  return (
    <SearchBarContainer>
      <Input
        type="text"
        value={value}
        onChange={(e: React.FormEvent<HTMLInputElement>) => handleChange(e, setInputValue)}
        onKeyPress={(e: React.KeyboardEvent<HTMLInputElement>) => handleKeyPress(e, userStore)}
        placeholder="Search for users"
      />
      <Button disabled={userStore.isLoading} onClick={() => userStore.fetchUserWithRepos(value)}>
        Search
      </Button>
    </SearchBarContainer>
  );
}

export default inject('userStore')(observer(SearchBar)) as any;
