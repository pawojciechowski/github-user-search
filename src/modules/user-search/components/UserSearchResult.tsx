import React from 'react';
import { observer, inject } from 'mobx-react';
import { IGithubUserStore } from '../stores/GithubUserStore';

interface UserSearchResultProps {
  user: IGithubUserStore
}

export function UserSearchResult({ user }: UserSearchResultProps) {
  return (
    <div>

    </div>
  )
}

export default inject('user')(observer(UserSearchResult)) as any;
