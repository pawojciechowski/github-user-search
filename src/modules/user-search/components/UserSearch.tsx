import React from 'react';
import styled from 'styled-components';
import SearchBar from './SearchBar';
import UserSearchResult from './UserSearchResult';
import Container from 'modules/themes/components/Container';
import { Theme } from 'modules/themes/types';

const UserSearchContainer = styled.div``;
const UserSearchHeader = styled(Container)`
  padding-top: ${({ theme }: { theme: Theme}) => `${theme.spacer * 1.5}px`};
  padding-bottom: ${({ theme }: { theme: Theme}) => `${theme.spacer * 1.5}px`};
  background: ${({ theme }: { theme: Theme}) => theme.foreground};
  box-shadow: ${({ theme }: { theme: Theme}) => theme.containerShadow};
`;
const UserSearchResultContainer = styled(Container)`
  padding-top: ${({ theme }: { theme: Theme}) => `${theme.spacer * 1.5}px`};
  padding-bottom: ${({ theme }: { theme: Theme}) => `${theme.spacer * 1.5}px`};
`;

function UserSearch() {
  return (
    <UserSearchContainer>
      <UserSearchHeader>
        <SearchBar />
      </UserSearchHeader>
      <UserSearchResultContainer>
        <UserSearchResult />
      </UserSearchResultContainer>
    </UserSearchContainer>
  )
}

export default UserSearch;
