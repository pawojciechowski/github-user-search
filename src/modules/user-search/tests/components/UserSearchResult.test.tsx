import React from 'react';
import { renderWithThemeProvider } from 'config/tests/utils';
import { UserSearchResult } from 'modules/user-search/components/UserSearchResult';
import { IGithubUserStore } from 'modules/user-search/stores/GithubUserStore';

describe('<UserSearchResult />', () => {
  it('should match snapshot when in initial state', () => {
    const userStore = ({
      state: 'initial',
    } as unknown) as IGithubUserStore;

    const { container } = renderWithThemeProvider(<UserSearchResult userStore={userStore} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when in notFound state', () => {
    const userStore = ({
      state: 'notFound',
      lastSearchUsername: 'testLastSearchUsername',
    } as unknown) as IGithubUserStore;

    const { container } = renderWithThemeProvider(<UserSearchResult userStore={userStore} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when in error state', () => {
    const userStore = ({
      state: 'error',
    } as unknown) as IGithubUserStore;

    const { container } = renderWithThemeProvider(<UserSearchResult userStore={userStore} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should match snapshot when in success state', () => {
    const userStore = ({
      state: 'success',
      userWithTopRepos: {
        login: 'testLogin',
        name: 'test Name',
        bio: 'testBio',
        avatarUrl: 'testUrl',
        repos: [
          {
            id: 1,
            name: 'testRepo1',
            htmlUrl: 'testRepoUrl1',
            stars: 1,
          },
        ],
      },
    } as unknown) as IGithubUserStore;

    const { container } = renderWithThemeProvider(<UserSearchResult userStore={userStore} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
