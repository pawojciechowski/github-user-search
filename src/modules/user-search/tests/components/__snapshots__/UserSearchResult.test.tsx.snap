// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserSearchResult /> should match snapshot when in error state 1`] = `
.c0 {
  font-size: test;
  font-weight: test;
  line-height: test;
  -webkit-letter-spacing: test;
  -moz-letter-spacing: test;
  -ms-letter-spacing: test;
  letter-spacing: test;
  color: test;
  margin-top: 0;
  margin-bottom: 1px;
  padding: 0;
}

<div
  class=""
>
  <h2
    class="c0"
  >
    There was an error while searching the user :(
  </h2>
</div>
`;

exports[`<UserSearchResult /> should match snapshot when in initial state 1`] = `
.c0 {
  font-size: test;
  font-weight: test;
  line-height: test;
  -webkit-letter-spacing: test;
  -moz-letter-spacing: test;
  -ms-letter-spacing: test;
  letter-spacing: test;
  color: test;
  margin-top: 0;
  margin-bottom: 1px;
  padding: 0;
}

.c1 {
  margin: 0;
  padding-bottom: 1px;
}

<div
  class=""
>
  <h2
    class="c0"
  >
    Let's find a Github user!
  </h2>
  <p
    class="c1"
  >
    Start with typing the username in the search box :)
  </p>
</div>
`;

exports[`<UserSearchResult /> should match snapshot when in loading state 1`] = `
.c3 {
  font-size: test;
  font-weight: test;
  line-height: test;
  -webkit-letter-spacing: test;
  -moz-letter-spacing: test;
  -ms-letter-spacing: test;
  letter-spacing: test;
  color: test;
  margin-top: 0;
  margin-bottom: 1px;
  padding: 0;
}

.c5 ul {
  list-style: none;
  padding: 0;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 li {
  display: block;
  margin-bottom: 1px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  margin-bottom: 1px;
}

.c0 .c2 {
  margin-bottom: 0;
}

.c1 {
  width: 64px;
  height: 64px;
  border-radius: 12px;
  overflow: hidden;
  margin-right: 1px;
}

.c4 {
  margin: 0;
  padding-bottom: 1px;
  margin-bottom: 1px;
}

<div
  class=""
>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <span>
        <span
          class="react-loading-skeleton css-1x45xix-skeletonStyles-Skeleton"
          style="height: 100%;"
        >
          ‌
        </span>
      </span>
    </div>
    <h1
      class="c2 c3"
    >
      <span>
        <span
          class="react-loading-skeleton css-1x45xix-skeletonStyles-Skeleton"
          style="width: 70px;"
        >
          ‌
        </span>
      </span>
      <br />
      <span>
        <span
          class="react-loading-skeleton css-1x45xix-skeletonStyles-Skeleton"
          style="width: 120px;"
        >
          ‌
        </span>
      </span>
    </h1>
  </div>
  <p
    class="sc-AxirZ c4"
  >
    <span>
      <span
        class="react-loading-skeleton css-1x45xix-skeletonStyles-Skeleton"
      >
        ‌
      </span>
      <span
        class="react-loading-skeleton css-1x45xix-skeletonStyles-Skeleton"
      >
        ‌
      </span>
    </span>
  </p>
  <div
    class="c5"
  >
    <h2
      class="c2 c3"
    >
      Top repositories
    </h2>
    <ul>
      <li>
        <span>
          <span
            class="react-loading-skeleton css-1x45xix-skeletonStyles-Skeleton"
            style="height: 48px;"
          >
            ‌
          </span>
        </span>
      </li>
      <li>
        <span>
          <span
            class="react-loading-skeleton css-1x45xix-skeletonStyles-Skeleton"
            style="height: 48px;"
          >
            ‌
          </span>
        </span>
      </li>
      <li>
        <span>
          <span
            class="react-loading-skeleton css-1x45xix-skeletonStyles-Skeleton"
            style="height: 48px;"
          >
            ‌
          </span>
        </span>
      </li>
    </ul>
  </div>
</div>
`;

exports[`<UserSearchResult /> should match snapshot when in notFound state 1`] = `
.c0 {
  font-size: test;
  font-weight: test;
  line-height: test;
  -webkit-letter-spacing: test;
  -moz-letter-spacing: test;
  -ms-letter-spacing: test;
  letter-spacing: test;
  color: test;
  margin-top: 0;
  margin-bottom: 1px;
  padding: 0;
}

.c1 {
  margin: 0;
  padding-bottom: 1px;
}

<div
  class=""
>
  <h2
    class="c0"
  >
    Couldn't find user with username "
    testLastSearchUsername
    " :(
  </h2>
  <p
    class="c1"
  >
    Please make sure you don't misspelled the username
  </p>
</div>
`;

exports[`<UserSearchResult /> should match snapshot when in success state 1`] = `
.c3 {
  font-size: test;
  font-weight: test;
  line-height: test;
  -webkit-letter-spacing: test;
  -moz-letter-spacing: test;
  -ms-letter-spacing: test;
  letter-spacing: test;
  color: test;
  margin-top: 0;
  margin-bottom: 1px;
  padding: 0;
}

.c5 ul {
  list-style: none;
  padding: 0;
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 li {
  display: block;
  margin-bottom: 1px;
}

.c6 {
  display: block;
  padding: 1px;
  background: test;
  box-shadow: test;
  border-radius: 12px;
  line-height: 16px;
  color: test;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  margin-bottom: 1px;
}

.c0 .c2 {
  margin-bottom: 0;
}

.c1 {
  width: 64px;
  height: 64px;
  border-radius: 12px;
  overflow: hidden;
  margin-right: 1px;
}

.c4 {
  margin: 0;
  padding-bottom: 1px;
  margin-bottom: 1px;
}

<div
  class=""
>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <img
        alt="Avatar url"
        src="testUrl"
      />
    </div>
    <h1
      class="c2 c3"
    >
      test
      <br />
      Name
      <br />
    </h1>
  </div>
  <p
    class="sc-AxirZ c4"
  >
    testBio
  </p>
  <div
    class="c5"
  >
    <h2
      class="c2 c3"
    >
      Top repositories
    </h2>
    <ul>
      <li>
        <a
          class="c6"
          href="testRepoUrl1"
          rel="noopener nofollow noreferer"
          target="_blank"
        >
          testRepo1
        </a>
      </li>
    </ul>
  </div>
</div>
`;
